bottlerocket-os/bottlerocket
# updog and migrator: signed migrations

## Issue number:

Closes #91
Closes #905 

## Description of changes:

Implements all changes necessary for migrator to load a locally cached tuf repo and obtain migrations from there. See #905 for a detailed explanation of the aigned migrations design.

Additionally, we decided to maintain compatibility so that instances can be upgraded into signed migrations. To facilitate this, in the current version, `upog` writes both signed and unsigned migrations. `migrator` checks the from version. If it's less than `0.3.5`, migrator executes unsigned migrations, otherwise it executes signed migrations.

In a future version, we can create a breaking change in which unsigned migrations are no-longer read. At that point, versions less than 0.3.5 must pass through a version that supports both signed and unsigned migrations (such as this one) in order to upgrade into signed migrations.

There is one behavioral change about signed migrations. Previously we used a `sort` of the migration names to provide stability to their order of execution (when more than one migration existed for a single version pair). With signed migrations, this is no longer necessary and the migrations will run in the order that they are listed in the manifest.

## Testing done:

The testing procedure is extensive. Note: a trace was added so that `migrator` announces in the system journal whether it is running signed or unsigned migrations. Also note, v0.3.1 was used as a starting point because a migration exists for v0.3.2 (which changes the host container version). Thus upgrading from v0.3.1 presents an actual unsigned migration.

### Test Setup
  
  * Create an AMI of v0.3.1, but with my non-production `root.json`.
  * Create matching image binaries for v0.3.1 with my non-production `root.json` and add them to a TUF repo (replacing the prod images).
  * Create a version of the code in this PR, naming the version 0.99.0, create an AMI of this and add the update binaries to the TUF repo/manifest.
  * Create a version of the code in this PR, along with a migration that adds a `foo` setting and sets its value to `bar`. Call this v0.99.1 and add it to the TUF repo/manifest.

### Test Execution

Starting from the v0.3.1 AMI, perform the following sequence of upgrades and downgrades.

  * Start at v0.3.1 via AMI
  * Upgrade v0.3.1 -> v0.99.0
  * Upgrade v0.99.0 -> v0.99.1
  * Downgrade v0.99.1 -> v0.99.0
  * Downgrade v0.99.0 -> v0.3.1

At each step along the way.

  * Ensure that Kubernetes is working by running a busybox pod.
  * Copy the system journal and Bottlerocket API settings to local storage.

At the end of the cycle, diff the system journals to observe that migrator annouced:

  * Upgrade v0.3.1 -> v0.99.0 'running unsigned migrations'
  * Upgrade v0.99.0 -> v0.99.1 'running signed migrations'
  * Downgrade v0.99.1 -> v0.99.0 'running signed migrations'
  * Downgrade v0.99.0 -> v0.3.1: older version of migrator made no announcement

Diff the API settings to observe the following setting changes:

  * Upgrade v0.3.1 -> v0.99.0: host container changed from v0.4.0 to v0.5.0
  * Upgrade v0.99.0 -> v0.99.1: "foo": "bar" was added
  * Downgrade v0.99.1 -> v0.99.0: "foo": "bar" was removed
  * Downgrade v0.99.0 -> v0.3.1: host container changed from v0.5.0 to v0.4.0

### Additional Testing

Additional testing has been performed with v0.99.0 AMI -> v0.99.1 -> v0.99.0, though this has not been repeated at every testing cycle.

## TODO

  * [x] Use `pentacle`
  * [x] Use `tough::RepoEditor` to create the `migrator` unit test repo on the fly.  ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  * [x] Use a single migration script for testing instead of two different migration scripts, and compress the migration script on the fly instead of checking in a compressed version of it. ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  * [x] Set repo expiration to 1970-01-01 to make it obvious that it is expired. Also make a note about this in the test. ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  
## Terms of contribution:

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
